src/model.jl

```diff
@@ -1,106 +1,275 @@
-function BatchNormWrap(out_ch)
-    Chain(x->expand_dims(x,2),
-	  BatchNorm(out_ch),
-	  x->squeeze(x))
-end
+using Flux
+using Flux: @functor
 
-UNetConvBlock(in_chs, out_chs, kernel = (3, 3)) =
-    Chain(Conv(kernel, in_chs=>out_chs,pad = (1, 1);init=_random_normal),
-	BatchNormWrap(out_chs),
-	x->leakyrelu.(x,0.2f0))
+struct ConvBlock
+  op
+end
 
-ConvDown(in_chs,out_chs,kernel = (4,4)) =
-  Chain(Conv(kernel,in_chs=>out_chs,pad=(1,1),stride=(2,2);init=_random_normal),
-	BatchNormWrap(out_chs),
-	x->leakyrelu.(x,0.2f0))
+@functor ConvBlock
 
-struct UNetUpBlock
-  upsample
+function ConvBlock(in_channels, out_channels, kernel_sizes = [(3,3), (3,3)];
+                       activation = NNlib.relu, padding = "valid")
+  pad_arg = padding == "same" ? SamePad() : 0
+  conv_layers = Any[]
+  in_channels_it = in_channels
+  for kernel_size in kernel_sizes
+    push!(conv_layers,
+      Conv(kernel_size, in_channels_it => out_channels, activation; pad=pad_arg)
+    )
+    in_channels_it = out_channels
+  end
+  return ConvBlock(Chain(conv_layers...))
 end
 
-@functor UNetUpBlock
+function (m::ConvBlock)(x)
+  return m.op(x)
+end
+  
+  # ConvBlock(in_chs, out_chs, kernel = (3, 3)) =
+  #   Chain(Conv(kernel, in_chs=>out_chs,pad = (1, 1);init=_random_normal),
+	# BatchNormWrap(out_chs),
+	# x->leakyrelu.(x,0.2f0))
 
-UNetUpBlock(in_chs::Int, out_chs::Int; kernel = (3, 3), p = 0.5f0) = 
-    UNetUpBlock(Chain(x->leakyrelu.(x,0.2f0),
-       		ConvTranspose((2, 2), in_chs=>out_chs,
-			stride=(2, 2);init=_random_normal),
-		BatchNormWrap(out_chs),
-		Dropout(p)))
+struct Downsample
+  op
+  factor
+  pooling_type
+end
 
-function (u::UNetUpBlock)(x, bridge)
-  x = u.upsample(x)
-  return cat(x, bridge, dims = 3)
+function Base.show(io::IO, d::Downsample)
+  print(io, "Unet.Downsample($(d.factor), $(d.pooling_type))")
 end
 
-"""
-    Unet(channels::Int = 1, labels::Int = channels)
+@functor Downsample
 
-  Initializes a [UNet](https://arxiv.org/pdf/1505.04597.pdf) instance with the given number of `channels`, typically equal to the number of channels in the input images.
-  `labels`, equal to the number of input channels by default, specifies the number of output channels.
-"""
-struct Unet
-  conv_down_blocks
-  conv_blocks
-  up_blocks
+function Downsample(downsample_factor; pooling_type="max")
```

Comment:
If we took the function `maxpool` or `meanpool` directly, we can get rid of the conditional.

Reply:
Good idea. Changed accordingly.